### PR TITLE
LRDOCS-8379 Introduction to REST Builder

### PR DESCRIPTION
**Pull request by Alec-Shay**
https://issues.liferay.com/browse/LRDOCS-8379
See: https://issues.liferay.com/browse/LPS-120387

This is the introduction to the concepts that are then used in the tutorial, which is also in review here: https://github.com/sez11a/liferay-learn/pull/151

There was quite a bit of information to introduce as to how REST Builder works, where everything goes, etc., so hopefully this didn't wind up being too wordy. Let me know what you think when you get the chance, though.

Also, one other thing to point out -- because this is adding a new article sitting beside the tutorial which is [also in review](https://github.com/sez11a/liferay-learn/pull/151), that means this commit (adding the new folder to the RSTs) is going to cause a conflict: 9591586

It's a pretty simple conflict, but if you'd like, I can either rebase all of this once that's merged, or just resolve the conflict in-place, or however you'd like me to address that. Just let me know. Thanks.

### CODE REVIEW

**Comment by javierdearcos** [comment]
Hello,

Great work synthesizing the big and difficult info needed to use REST Builder.

I've just finished doing the SME review. I'm going to add some minor feedback if you consider it may help. 

I'll add it as a comment in the concrete line but please, tell me what it is the best approach for you to use in the future.

Thank you!

File: docs/dxp/7.x/en/headless-delivery/content-delivery-apis/rest-builder/introduction-to-rest-builder.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+See the [OpenAPI specification](https://swagger.io/docs/specification/data-models/data-types/) for a list of all the supported basic data types that you can use for schemas.
+
+```tip::
+   If your API only needs to return one of the basic data types, then you may not need to define any schemas in your ``rest-openapi.yaml`` file.
+```
+
+### Defining the APIs
+
+The last section, `paths`, specifies the details of the APIs themselves. Each API must be defined within its own path, which is added to the end of the URL to access it when it is available. Making a request to the full URL including the path will call the API you define within it. APIs can be defined to take either `get` or `post` requests.
```
Comment (javierdearcos) [inline docs/dxp/7.x/en/headless-delivery/content-delivery-apis/rest-builder/introduction-to-rest-builder.md:137]: APIs can be defined to use get, post, put, patch and delete

File: docs/dxp/7.x/en/headless-delivery/content-delivery-apis/rest-builder/introduction-to-rest-builder.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+REST Builder generates several packages of code for you, so that you can focus on just your implementation class. However, you must start by creating the locations for each of your modules. In a conventional Liferay workspace, you only need to create the folders for each module and each of their `bnd.bnd` and `build.gradle` files.
+
+After you create the initial configuration for your modules,  and the [REST Builder configuration files](#adding-your-api-configurations) (`rest-config.yaml` and `rest-openapi.yaml`), then running REST Builder generates the following structure:
+
+```
+api root
+├── bnd.bnd
+├── build.gradle
+└── <apiDir>
```
Comment (javierdearcos) [inline docs/dxp/7.x/en/headless-delivery/content-delivery-apis/rest-builder/introduction-to-rest-builder.md:23]: I think is clearer using the src/main/java reference as in the impl structure instead of the <apiDir>

File: docs/dxp/7.x/en/headless-delivery/content-delivery-apis/rest-builder/introduction-to-rest-builder.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                            schema:
+                                $ref: "#/components/schemas/BasicUser"
+                    description:
+                        "The BasicUser object"
+            tags: ["BasicUser"]
+```
+
+The `parameters` each have a `name` (which is used both in specifying it in a request as well as in the Java method call), whether it is required, and a `schema` (which may be either a basic data type or a schema for a custom object you have defined).
+
+The `responses` each need to have `content` defined for both JSON and XML responses. Each of the possible responses must also define a `schema`, just like the parameters.
```
Comment (javierdearcos) [inline docs/dxp/7.x/en/headless-delivery/content-delivery-apis/rest-builder/introduction-to-rest-builder.md:168]: Responses can have content only defined in JSON or XML format. Although the other format can be autogenerated depending on the configuration used in the rest-config file

**Review by javierdearcos — COMMENTED** [review]
SME review

**Comment by Alec-Shay** [comment]
Thank you @javierdearcos! To answer your question, I think this format works great, since it removes any uncertainty as to what part of the article you're talking about with your feedback.

@sez11a I just pushed a new commit incorporating these improvements

**Review by javierdearcos — APPROVED** [review]
I think is perfect now. Great work! Thank you!

**Comment by sez11a** [comment]
Just started reviewing :)

**Comment by sez11a** [comment]
Superseded by https://github.com/jrhoun/liferay-learn/pull/678. Thanks, @Alec-Shay and @javierdearcos !
